### ISSUE TITLE
Trying to chain in_string() with concat. Bug ?

### ISSUE THREAD

**Issue by sol-prog** [nfriend/cooltojs#9]
Hello,

When I try to run this code:

```
class Main inherits IO {
    main() : Object {
            out_string(in_string().concat("\n"))
    };
};
```

I get a runtime error:

```
Transpilation successful!
TypeError: this.in_string(...).concat is not a function
```
The code is directly taken from course. (I've tried the code in the provided VM and it works).
I'm curious if this is a bug in your compiler to TypeScript. BTW, cool project. Really helpful.

If I use an intermediary variable it works as expected:

```
class Main inherits IO {
    inp: String;
    
    main() : Object {
        {
            inp <- in_string(); 
            out_string(inp.concat("\n"));
        }
    };
};
```

Thanks

**Comment by nfriend** [comment]
Hi @sol-prog - thanks for the report!  I took a look, and you're right, this is a bug in my compiler.  The compiler is translating your COOL code into this ES6 JavaScript:

```js
_returnValue = this.out_string(yield* this.in_string(_inputGenerator).concat(new _String("\n")))
```

What it _should_ be doing is adding an extra pair of parenthesis around the `yield*` expression:

```js
                               ↓                                      ↓
_returnValue = this.out_string((yield* this.in_string(_inputGenerator)).concat(new _String("\n")))
```

That way the `.concat()` function isn't called until the generator returns.

I might try and find the root cause of this to see if it's a quick fix.  But no promises - it's been several years since I lasted cracked open this codebase :).
